servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +179,125 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
+                                    new RequestTimeoutSubscriber(subscriber,
+                                            SourceAdapters.fromSource(requestProcessor), timeout, timeoutExecutor))
                             .shareContextOnSubscribe();
                 });
             }
         };
     }
 
+    // package private for testing purposes
+    static final class RequestTimeoutSubscriber implements Subscriber<StreamingHttpResponse> {
+
+        private static final AtomicIntegerFieldUpdater<RequestTimeoutSubscriber> onceUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(RequestTimeoutSubscriber.class, "once");
+
+        private final DelayedCancellable requestCancellable = new DelayedCancellable();
+        private final Cancellable timeoutCancellable;
+        private final Subscriber<? super StreamingHttpResponse> delegate;
+
+        private final Duration timeout;
+        private final Executor timeoutExecutor;
+        @SuppressWarnings("unused")
+        private volatile int once;
+
+        RequestTimeoutSubscriber(Subscriber<? super StreamingHttpResponse> delegate, Completable requestComplete,
+                                 Duration timeout, Executor timeoutExecutor) {
+            this.delegate = delegate;
+            this.timeout = timeout;
+            this.timeoutExecutor = timeoutExecutor;
+            timeoutCancellable = requestComplete.concat(Completable.never()
+                    .timeout(timeout, timeoutExecutor)).beforeOnError(this::handleInterruptions).subscribe();
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            try {
+                delegate.onSubscribe(() -> {
+                    once();
+                    Throwable t = null;
+                    try {
+                        timeoutCancellable.cancel();
+                    } catch (Throwable tt) {
+                        t = tt;
+                    }
+                    try {
+                        requestCancellable.cancel();
+                    } catch (Throwable tt) {
+                        if (t == null) {
+                            t = tt;
+                        } else {
+                            t.addSuppressed(tt);
+                        }
+                    }
+                    if (t != null) {
+                        throwException(t);
+                    }
+                });
+                requestCancellable.delayedCancellable(cancellable);
+            } catch (Throwable cause) {
+                try {
+                    cancellable.cancel();
+                } catch (Throwable t) {
+                    cause.addSuppressed(t);
+                }
+                onError(cause);
+                LOGGER.warn("Unexpected exception from onSubscribe of Subscriber {}.", delegate, cause);
+            }
+        }
+
+        private void handleInterruptions(Throwable t) {
+            if (once()) {
+                requestCancellable.cancel();
```

Comment:
If you decide to keep `try-catch` around `timeoutCancellable.cancel()` in `onError` and `onSuccess`, we also need one here.

Reply:
Based on the chat discussion: it looks like we need to try-catch this call because users may add filters that will intercept this call and as the result, we risk to encounter an exception in users code